File: src/algorithms/linked-lists/reverse-linked-list/sources/reverse-linked-list.py
# Reverse Linked List — iteratively redirect each node's next pointer to its predecessor
class ListNode:
    def __init__(self, value, next_node=None):
        self.value = value
        self.next = next_node


def reverse_linked_list(head):
    prev = None  # @step:initialize
    current = head  # @step:initialize
    while current is not None:
        next_node = current.next  # @step:traverse-next
        current.next = prev  # @step:reverse-pointer
        prev = current  # @step:reverse-pointer
        current = next_node  # @step:traverse-next
    return prev  # @step:complete

File: src/algorithms/linked-lists/manipulation/reverse-linked-list/sources/ReverseLinkedList.cpp
// Reverse Linked List — iteratively redirect each node's next pointer to its predecessor
struct ListNode {
    int value;
    ListNode* next;
    ListNode(int val) : value(val), next(nullptr) {}
};

ListNode* reverseLinkedList(ListNode* head) {
    ListNode* prev = nullptr; // @step:initialize
    ListNode* current = head; // @step:initialize
    while (current != nullptr) {
        ListNode* nextNode = current->next; // @step:traverse-next
        current->next = prev; // @step:reverse-pointer
        prev = current; // @step:reverse-pointer
        current = nextNode; // @step:traverse-next
    }
    return prev; // @step:complete
}
